src/ast_editor.ml

```diff
@@ -0,0 +1,514 @@
+            Int.of_string (Ojs.string_of_js (Ojs.get_prop_ascii msg "r_begin"))
+          in
+          let rcend =
+            Int.of_string (Ojs.string_of_js (Ojs.get_prop_ascii msg "r_end"))
+          in
+          apply_selection editor rcbegin rcend
+        | _ -> ())
+      maybe_pair_editors
+  else
+    ()
+
+let on_hover custom_doc webview =
+  let hover =
+    Hover.make ~contents:(`MarkdownString (MarkdownString.make ~value:"" ()))
+  in
+  let provideHover ~(document : TextDocument.t) ~(position : Position.t)
+      ~token:_ =
+    let offset = TextDocument.offsetAt document ~position in
+    if document_eq custom_doc document then
+      send_msg "focus" (Ojs.int_to_js offset) ~webview
+    else
+      ();
+    `Value (Some [ hover ])
+  in
+  let provider = HoverProvider.create ~provideHover in
+  Vscode.Languages.registerHoverProvider ~selector:(`String "ocaml") ~provider
+
+let activate_hover_mode instance ~document =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match Ast_editor_state.find_webview_by_doc ast_editor_state ~document with
+  | Some webview -> on_hover document webview
+  | None -> failwith "Webview wasn't found while switching hover mode"
+
+let resolveCustomTextEditor instance ~(document : TextDocument.t) ~webviewPanel
+    ~token:_ : CustomTextEditorProvider.ResolvedEditor.t =
+  let webview = WebviewPanel.webview webviewPanel in
+  (*persist the webview*)
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  Ast_editor_state.set_webview_map ast_editor_state
+    (Map.set
+       (Ast_editor_state.get_webview_map ast_editor_state)
+       ~key:(doc_string_uri ~document) ~data:webview);
+  let onDidChangeTextDocument_disposable =
+    Workspace.onDidChangeTextDocument
+      ~listener:(onDidChangeTextDocument_listener ~webview ~document)
+      ()
+  in
+  let onDidReceiveMessage_disposable =
+    WebView.onDidReceiveMessage webview
+      ~listener:(onDidReceiveMessage_listener instance ~document)
+      ()
+  in
+  let (_ : Disposable.t) =
+    WebviewPanel.onDidDispose webviewPanel
+      ~listener:(fun () ->
+        Ast_editor_state.set_original_mode ast_editor_state true;
+        Disposable.dispose onDidReceiveMessage_disposable;
+        Disposable.dispose onDidChangeTextDocument_disposable)
+      ()
+  in
+  transform_to_ast ~document ~webview;
+  let options = WebView.options webview in
+  WebviewOptions.set_enableScripts options true;
+  WebView.set_options webview options;
+  let p =
+    let open Promise.Syntax in
+    let+ r = read_html_file () in
+    WebView.set_html webview r
+  in
+  `Promise p
+
+let open_ast_explorer ~uri =
+  let open Promise.Syntax in
+  let+ _ =
+    Vscode.Commands.executeCommand ~command:"vscode.openWith"
+      ~args:
+        [ Uri.t_to_js uri
+        ; Ojs.string_to_js "ast-editor"
+        ; ViewColumn.t_to_js ViewColumn.Beside
+        ]
+  in
+  ()
+
+let replace_document_content ~document ~content =
```

Comment:
I don't see the return value of this function used anywhere. Therefore it should return just return unit

Reply:
Changed in d17892794eb016a5a250c82484080217ed54c802